### PR TITLE
Fix dropped subscriptions due to ETH RPC client reconnects

### PR DESCRIPTION
**Pull request by yondonfu**
A recently observed issue on mainnet is that when a node waits for a transaction to be mined, if the transaction takes too long to be mined the node will get a `context deadline exceeded` error message and when the node later makes another RPC call oftentimes it will stop receiving subscription messages (i.e. new blocks, new jobs, etc.) from the ETH node that it is connected to. #455 is one such case.

This behavior is caused by a race condition created by the `WaitMined()` helper - the function accepts a context which is not only used to signal the end of the polling loop, but is also passed in to the `TransactionReceipt()` function (https://github.com/livepeer/go-livepeer/blob/master/vendor/github.com/ethereum/go-ethereum/accounts/abi/bind/util.go#L37) which will execute an RPC call to attempt to fetch the transaction receipt if it exists. Common usage of this helper would involve passing in a context with a timeout. But the context is also used by the RPC call within the polling loop which wouldn't normally be a problem except for the fact that if the deadline for the context used by the RPC call is exceeded then the RPC client will treat the timed out network write as a signal to establish a new connection on a subsequent RPC call (https://github.com/livepeer/go-livepeer/blob/master/vendor/github.com/ethereum/go-ethereum/rpc/client.go#L491). Furthermore, when the client tries to establish a new connection on a subsequent RPC call, the existing connection will be closed (https://github.com/livepeer/go-livepeer/blob/master/vendor/github.com/ethereum/go-ethereum/rpc/client.go#L577). Closing the existing connection will tell the ETH node to stop sending notifications for all subscriptions created using the connection (https://github.com/ethereum/go-ethereum/blob/master/eth/filters/api.go#L227).

I believe the race condition is triggered whenever the context deadline is exceeded when the network write for `TransactionReceipt()` has already been initiated - if the context deadline is exceeded before the network write for `TransactionReceipt()` is initiated I don't think the race condition is triggered.

This PR proposes a small change to mitigate the described issue by checking if a network write error is due to a timeout and if so the RPC client will not try to establish a new connection (and close the existing one) during a subsequent RPC call - instead it will continue to try to use the same existing connection. If the network write error is not due to a timeout (i.e. an EOF), it will try to establish a new connection during a subsequent RPC call.

### CODE REVIEW

**Comment by j0sh** [comment]
Nice diving.

Seems the root of the issue is that the passed-in context should not have been shared with the RPC call, since they're two different things: waiting for the transaction to be mined, and the actual RPC call checking whether the transaction exists. If there is a legitimate network error while performing the RPC call, then it stands to reason that other subscriptions depending on the connection would be affected. However, if the `WaitMined` timeout occurs mid-RPC, that should not have an effect on other subscriptions. The best course of action here seems to be to allow the RPC call to finish, then check whether the `WaitMined` timeout has elapsed.

I'm also a bit concerned with broadly special-casing RPC errors for network timeout, since that may mask other legitimate issues. Also, since we already take pains to retry failed RPC connections, would this change introduce further side effects with a legitimate network timeout? Would it affect Infura?

An alternate solution could be to create a new context for each iteration of the RPC call in `WaitMined` . Note that it's easy to inadvertently re-introduce the same bug this way, which could happen by decreasing the context deadline based on time elapsed, or by using the `WaitMined` ctx as the parent context. So I'd keep it simple.

Lastly, a small note on procedure -- this repo is a better place for discussion, but before merging anything, could the changes be put in https://github.com/livepeer/go-ethereum with the commit hash reflected in the commit message here? For an example from LPMS: https://github.com/livepeer/go-livepeer/commit/f2bbdbf074a0e4cc1f01b520faec43214ac821b3

**Comment by yondonfu** [comment]
> Seems the root of the issue is that the passed-in context should not have been shared with the RPC call

While this is related to why the problem arises, I'm not sure if this is necessarily wrong since passing in a context into a long running function which then either passes in the same context or a derived child context to other nested functions that should be cancellable is a common pattern in Go. This particular case ends up being problematic for two reasons. 1. The context passed into the RPC call is not only used to pass down a cancellation signal, but also used to define the deadline for the network write (https://github.com/ethereum/go-ethereum/blob/master/rpc/client.go#L488). If the parent context deadline is exceeded, but the network write completes then the in-progress RPC call can still be cancelled while the client is waiting for the actual response from the network write. However, if the parent context deadline is exceeded and the network write is in-progress then the network write times out since it shares the same deadline as the parent context. 2. The RPC client will close the existing connection and open a new one if it ever gets a network write timeout. I'm not sure if this is always ideal since the user controls the timeout and if it ever passes in a low value timeout, the existing connection will be closed even if there might not have been anything wrong with that connection. In this case, each RPC call in the polling loop basically gets a smaller and smaller timeout because the same deadline is used for the network write for each RPC call. 

> Also, since we already take pains to retry failed RPC connections, would this change introduce further side effects with a legitimate network timeout? Would it affect Infura?

We haven't run into any timeout related issues with Infura as of late AFAIK - the connection issues are usually due to Infura closing the connection thereby sending an EOF to the client. However,  I agree that it is a good point to be wary of special casing network timeout errors - I'm not sure about the side effects.

> An alternate solution could be to create a new context for each iteration of the RPC call in WaitMined

I thought to do this originally, but hesitated because we wouldn't be able to derive the context from the parent context passed into `WaitMined()` because if the parent context has a deadline attached the child context will have the same deadline which results in the same problem. In order for this to work we would have to define a new independent context derived from `context.Background()` or use `context.Background()` directly which kind of goes against the advised use of the context API to have contexts flow through function calls in a tree form and `context.Background()` only being used in the main function, initialization, tests and for the top-level context in an incoming network requests. Then again, I don't think we adhere to this advised usage at the moment anyway so perhaps it is ok for now.

We could change `WaitMined()` to

```
func WaitMined() {
    ...
    localCtx, cancel := context.WithCancel(context.Background())
    defer cancel()

    for {
        receipt, err := b.TransactionReceipt(localCtx, tx.Hash()
        ...
    }
}
```

The newly created local context would just be cancelled whenever the function exits which would happen when the parent context passed into `WaitMined()` is cancelled, likely via a timeout deadline. Since this local context does not have a deadline, the RPC client will use a default network write deadline instead https://github.com/ethereum/go-ethereum/blob/master/rpc/client.go#L480

Let me know what you think about this.

> Lastly, a small note on procedure -- this repo is a better place for discussion, but before merging anything, could the changes be put in https://github.com/livepeer/go-ethereum with the commit hash reflected in the commit message here?

Yep, I can do that after we wrap up the discussion here.

**Comment by j0sh** [comment]
> passing in a context into a long running function which then either passes in the same context or a derived child context to other nested functions that should be cancellable is a common pattern in Go

Yeah, I just think the pattern breaks down here for the reasons you outlined. Separate contexts seem more appropriate. (Or perhaps not cancelling unrelated subscriptions, but that's a much larger change.)

> localCtx, cancel := context.WithCancel(context.Background())

Looks good to me. `TransactionReceipt` is blocking, right? There's no way it'd be preempted by the passed-in `WaitMined` context expiring?

**Comment by yondonfu** [comment]
> Looks good to me. TransactionReceipt is blocking, right? There's no way it'd be preempted by the passed-in WaitMined context expiring?

Yeah `TransactionReceipt()` should block - just pushed the update for it to use its own context. Also made the relevant change in https://github.com/livepeer/go-ethereum and references the commit hash

**Review by j0sh — APPROVED** [review]
LGTM 🚢

**Pull request merged**
